drivers/ethos/ethos.c

```diff
@@ -299,38 +294,112 @@ static void _get_mac_addr(netdev_t *encdev, uint8_t* buf)
     memcpy(buf, dev->mac_addr, 6);
 }
 
+static unsigned _copy_byte(uint8_t *buf, uint8_t byte, bool *escaped)
+{
+    *buf = byte;
+    *escaped = false;
+    return 1U;
+}
+
+unsigned ethos_unstuff_readbyte(uint8_t *buf, uint8_t byte, bool *escaped, uint8_t *frametype)
+{
+    unsigned res = 0U;
+
+    switch (byte) {
+        case ETHOS_ESC_CHAR:
+            *escaped = true;
+            /* Intentionally falls through */
+        case ETHOS_FRAME_DELIMITER:
+            break;
+        case ETHOS_ESC_CHAR ^ 0x20:
+            if (*escaped) {
+                return _copy_byte(buf, ETHOS_ESC_CHAR, escaped);
+            }
+            /* Intentionally falls through */
+            /* to default when !(*escaped) */
+        case ETHOS_FRAME_DELIMITER ^ 0x20:
+            if (*escaped) {
+                return _copy_byte(buf, ETHOS_FRAME_DELIMITER, escaped);
+            }
+            /* Intentionally falls through */
+            /* to default when !(*escaped) */
+        case ETHOS_FRAME_TYPE_TEXT ^ 0x20:
+        case ETHOS_FRAME_TYPE_HELLO ^ 0x20:
+        case ETHOS_FRAME_TYPE_HELLO_REPLY ^ 0x20:
+            if (*escaped) {
+                *frametype = byte ^ 0x20;
+                break;
+            }
+            /* Intentionally falls through */
+        default:
+            return _copy_byte(buf, byte, escaped);
+    }
+    return res;
+}
+
 static int _recv(netdev_t *netdev, void *buf, size_t len, void* info)
 {
     (void) info;
     ethos_t * dev = (ethos_t *) netdev;
+    int res = 0;
 
     if (buf) {
-        if (len < dev->last_framesize) {
-            DEBUG("ethos _recv(): receive buffer too small.\n");
-            return -1;
-        }
+        int byte;
+        bool escaped = false;
+        uint8_t *ptr = buf;
+        uint8_t frametype = ETHOS_FRAME_TYPE_DATA;
 
-        len = dev->last_framesize;
+        do {
+            int tmp;
 
-        if ((tsrb_get(&dev->inbuf, buf, len) != (int)len)) {
-            DEBUG("ethos _recv(): inbuf doesn't contain enough bytes.\n");
-            dev->last_framesize = 0;
-            return -1;
-        }
+            if ((byte = tsrb_get_one(&dev->inbuf)) < 0) {
+                DEBUG("ethos _recv(): inbuf doesn't contain enough bytes.\n");
+                return -EIO;
+            }
+            tmp = ethos_unstuff_readbyte(ptr, (uint8_t)byte, &escaped, &frametype);
```

Comment:
I don't understand the escaping yet (stay with me, I think I start understanding), but maybe you can clarify this easily: If a stdio_read() is started between having received an escape and the escaped data, will this not produce an erroneous read?

Say the buffer is filled with a text frame, and another frame (that won't fit in the buffer yet) is being filled in. When the buffer is full, `stdio_read` is called and (with speed varying depending on which high-priority threads are running) pulls out data byte by byte, creating on and off opportunity for isrpipe_write_one to push data.

Frametype stays text, data keeps flowing in and getting dropped on and off before the reader code even gets to the being-corrupted frame, and all that happens on isrpipe_write_one failure is that dev->inbuf gets touched (which is weird as I'd understand that to be network related).

The more I think of this, the more it seems to me that the `puts("lost frame");` code in the ETHOS_FRAME_TYPE_TEXT case doesn't do what it should. Maybe it should set the frame type to FRAME_TYPE_ERRORED and thus ensure that the rest of the frame is discarded (rather than risking that individual characters get eaten and possibly misassembled at the stdio reading site)?

(Not sure of any of this, but it seems weird; gotta cut it short now and will continue at a later point.)

Reply:
Not 100% sure I understood your scenario, so let me just describe how the buffers and escaping work:

For data, hello and hello reply packets, the frame-type is determined by the head, consisting of the escaped frame type or just (escaped) data, when it is data which is pushed to the `tsrb` with the remainder of the packet. Also: at the end of the frame there is a frame delimiter character (if it appears in the frame it is escaped). So if multiple frames are in the ring buffer, one frame will be read until the frame delimiter (and the type determined by its head) and the same is true for the next.

Text frames (which in the ISR arrive in sequence with data frames) are not put to the `tsrb` ring buffer, but to the `isrpipe`. Since only text frames go there (and since writing back the escape + type seemed to be more trouble than its worth) they are not identified to the handler thread with a head, but are delimited by the frame delimiter.

In both buffers, the data is escaped similar to SLIP, so if the frame delimiter or the escape byte appear within the packet, they are escaped with the byte sequence, `{escape_byte, c ^ 0x20}`.

Packet types (at least as far as I understand it) are only escaped at the beginning of the packet. This is where the protocol design is somewhat flawed, because it seems a bit like a "because we can approach" and also makes it necessary to do at least some rudimentary unstuffing to determine into which buffer (data or text) the frame should go.

Reply:
Thanks for the input, with that I'd try to make the point more precise:

* If overflow happens in _handle_char for Ethernet frames, the whole frame is dropped (that's how I understand the `inbuf.{reads,writes} - 0`), and thanks to the frametype being reset, further data is ignored (is it? `->frametype` is also reset, but we're still in the `case IN_FRAME` of ethos_isr, so wouldn't the next byte trigger the default case assert?). Dropping a frame is relatively safe (a counter might be nice but adding features is not what this PR is about).
* If overflow happens in _handle_char on text frames, the network frame is dropped (?; we're not in one), but otherwise the error is ignored.

  As text may be read piecemeal, we can't "cancel" the whole text frame (some of it may have been consumed, that's OK), but IMO the sane behavior would be to discard the rest of the frame -- for the alternative is that while more bytes come in, some may be accepted and some not, and the ISR (rightfully, after this PR) feeds all the characters until end of frame to the isrpipe. For example, if the data stream is "H e l l o Escape EscapedEscape", and the Escape gets dropped but the EscapedEscape does not, the stdio reader would not read the bytes "H e l l o Escape" (which there should be) but "H e l l o EscapedEscape".

  I think that on the UART, losing text is OK -- receiving wrong text is not. I have a slight preference for losing chunks of text over losing individual characters, but (now that escapes are decoded after the isrpipe) losing individual characters has the downside of possibly producing wrong text, and thus, losing a chunk (the rest of the text frame, ) it should be.

Reply:
I think I start to understand what you mean. Will fix.